torch/nn/modules/_functions.py

```diff
@@ -0,0 +1,90 @@
+import torch
+from torch.autograd.function import Function
+
+
+class SyncBatchNorm(Function):
+
+    @staticmethod
+    def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
+        input = input.contiguous()
+
+        # calcualte mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+
+        mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
+        invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
+        mean_l = list(mean_all.unbind(0))
+        invstd_l = list(invstd_all.unbind(0))
+        # using all_gather instead of all reduce so we can calculate mean/var in one go
+        torch.distributed.all_gather(mean_l, mean, process_group)
+        torch.distributed.all_gather(invstd_l, invstd, process_group)
+
+        # calcualte global mean & invstd
+        mean, invstd = torch.batch_norm_update_stats(
+            input,
+            mean_all,
+            invstd_all,
+            running_mean,
+            running_var,
+            momentum,
+            eps,
+            int(input.numel() / input.size(1))
+        )
+
+        self.save_for_backward(input, weight, mean, invstd)
+        self.process_group = process_group
+        self.world_size = world_size
+
+        # apply element-wise normalization
+        out = torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+        return out
+
+    @staticmethod
+    def backward(self, grad_output):
+        grad_output = grad_output.contiguous()
+        saved_input, weight, mean, invstd = self.saved_tensors
+        grad_input = grad_weight = grad_bias = None
+        process_group = self.process_group
+        world_size = self.world_size
+
+        # calculate local stats as well as grad_weight / grad_bias
+        mean_dy, mean_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(
+            grad_output,
+            saved_input,
+            mean,
+            invstd,
+            self.needs_input_grad[0],
+            self.needs_input_grad[1],
+            self.needs_input_grad[2]
+        )
+
+        if self.needs_input_grad[0]:
+            # synchronizing stats used to calculate input gradient.
+            # TODO: move div_ into batch_norm_backward_elemt kernel
+            if torch.distributed.is_initialized():
```

Comment:
That's a pretty flaky check for being used within DDP. It's not the only case when people use distributed, and it might cause deadlocks for people.

Also, we should consider doing those asynchronously if possible, so we could unblock other branches of the backwards pass.

Reply:
I should have removed `if torch.distributed.is_initialized():` here, since I was doing the check in `SyncBatchNorm` module already.

So the follow up question is, what should be the proper way to check (here https://github.com/pytorch/pytorch/pull/14267/files#diff-98c9603e09a23834975e9d62c07c30d9R427) whether distributed is used?

I didn't quite get 
>  we should consider doing those asynchronously if possible
Are you referring to the all_reduce/gather here that should be done asynchronously?

Reply:
> Are you referring to the all_reduce/gather here that should be done asynchronously?

Yes, the new distributed module can return futures, but after giving it some though it's probably not doable since you have to have valid CUDA memory for next kernels at this point.

> So the follow up question is, what should be the proper way to check (...) whether distributed is used?

From what I understand this module is supposed to be used from within `DistributeDataParallel` only. You could define a thread-local Python variable that is set when `DistributedDataParallel.forward` executes, and is unset otherwise. Then, check it inside this function.